resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,575 @@
+	 */
+	String getName();
+
+	/**
+	 * Returns the AdaptiveBulkheadConfig of this Bulkhead.
+	 *
+	 * @return bulkhead config
+	 */
+	AdaptiveBulkheadConfig getBulkheadConfig();
+
+	/**
+	 * Get the AdaptiveMetrics of this Bulkhead.
+	 *
+	 * @return the AdaptiveMetrics of this Bulkhead
+	 */
+	AdaptiveMetrics getMetrics();
+
+	/**
+	 * Returns an EventPublisher which subscribes to the reactive stream of BulkheadEvent/BulkheadLimit events and
+	 * can be used to register event consumers.
+	 *
+	 * @return an AdaptiveEventPublisher
+	 */
+	AdaptiveEventPublisher getEventPublisher();
+
+	/**
+	 * Decorates and executes the decorated Supplier.
+	 *
+	 * @param supplier the original Supplier
+	 * @param <T>      the type of results supplied by this supplier
+	 * @return the result of the decorated Supplier.
+	 */
+	default <T> T executeSupplier(Supplier<T> supplier) {
+		return decorateSupplier(this, supplier).get();
+	}
+
+	/**
+	 * Decorates and executes the decorated Callable.
+	 *
+	 * @param callable the original Callable
+	 * @param <T>      the result type of callable
+	 * @return the result of the decorated Callable.
+	 * @throws Exception if unable to compute a result
+	 */
+	default <T> T executeCallable(Callable<T> callable) throws Exception {
+		return decorateCallable(this, callable).call();
+	}
+
+	/**
+	 * Decorates and executes the decorated Runnable.
+	 *
+	 * @param runnable the original Runnable
+	 */
+	default void executeRunnable(Runnable runnable) {
+		decorateRunnable(this, runnable).run();
+	}
+
+	/**
+	 * Decorates and executes the decorated Supplier.
+	 *
+	 * @param checkedSupplier the original Supplier
+	 * @param <T>             the type of results supplied by this supplier
+	 * @return the result of the decorated Supplier.
+	 * @throws Throwable if something goes wrong applying this function to the given arguments
+	 */
+	default <T> T executeCheckedSupplier(CheckedFunction0<T> checkedSupplier) throws Throwable {
+		return decorateCheckedSupplier(this, checkedSupplier).apply();
+	}
+
+	/**
+	 * Decorates and executes the decorated CompletionStage.
+	 *
+	 * @param supplier the original CompletionStage
+	 * @param <T>      the type of results supplied by this supplier
+	 * @return the decorated CompletionStage.
+	 */
+	default <T> CompletionStage<T> executeCompletionStage(Supplier<CompletionStage<T>> supplier) {
+		return decorateCompletionStage(this, supplier).get();
+	}
+
+	/**
+	 * Returns a supplier which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the Bulkhead
+	 * @param supplier the original supplier
+	 * @param <T>      the type of results supplied by this supplier
+	 * @return a supplier which is decorated by a Bulkhead.
+	 */
+	static <T> CheckedFunction0<T> decorateCheckedSupplier(AdaptiveBulkhead bulkhead, CheckedFunction0<T> supplier) {
+		return () -> {
+			long start = 0;
+			boolean isFailed=false;
+			bulkhead.acquirePermission();
+			try {
+				start = Instant.now().toEpochMilli();
```

Comment:
Can we use everywhere `final long start = System.nanoTime();`.
It reduces the amount of `Instant` classes which we create per call and improves GC.

Reply:
I feel it is like more optimized in current impl , as the allocation of start to now will only happen IF bulkhead.acquirePermission() pass plus we need the reference of start time in the finally block .